src/apps/vpn/connectionmanager.cpp

```diff
@@ -0,0 +1,999 @@
+                !m_serverData.entryServerPublicKey().isEmpty()
+            ? MozillaVPN::instance()->serverCountryModel()->server(
+                  m_serverData.entryServerPublicKey())
+            : Server::weightChooser(m_serverData.entryServers());
+
+    if (!entryServer.initialized()) {
+      logger.error() << "Empty entry server list in state" << m_state;
+      serverUnavailable();
+      return;
+    }
+
+    InterfaceConfig entryConfig;
+    entryConfig.m_privateKey = vpn->keys()->privateKey();
+    entryConfig.m_deviceIpv4Address = device->ipv4Address();
+    entryConfig.m_deviceIpv6Address = device->ipv6Address();
+    entryConfig.m_serverPublicKey = entryServer.publicKey();
+    entryConfig.m_serverIpv4AddrIn = entryServer.ipv4AddrIn();
+    entryConfig.m_serverIpv6AddrIn = entryServer.ipv6AddrIn();
+    entryConfig.m_serverPort = entryServer.choosePort();
+    entryConfig.m_hopType = InterfaceConfig::MultiHopEntry;
+    entryConfig.m_allowedIPAddressRanges.append(
+        IPAddress(exitServer.ipv4AddrIn()));
+    entryConfig.m_allowedIPAddressRanges.append(
+        IPAddress(exitServer.ipv6AddrIn()));
+
+    // If requested, force the use of port 53/DNS.
+    if (dnsPort == ForceDNSPort) {
+      logger.info() << "Forcing port 53";
+      entryConfig.m_serverPort = 53;
+    }
+
+    m_activationQueue.append(entryConfig);
+  }
+  // Otherwise, we can approximate multihop support by redirecting the
+  // connection to the exit server via the multihop port.
+  else {
+    logger.info() << "Activating multi-hop (through multihop port)";
+    exitConfig.m_hopType = InterfaceConfig::SingleHop;
+
+    Server entryServer =
+        serverSelectionPolicy == DoNotRandomizeServerSelection &&
+                !m_serverData.entryServerPublicKey().isEmpty()
+            ? MozillaVPN::instance()->serverCountryModel()->server(
+                  m_serverData.entryServerPublicKey())
+            : Server::weightChooser(m_serverData.entryServers());
+
+    if (!entryServer.initialized()) {
+      logger.error() << "Empty entry server list in state" << m_state;
+      serverUnavailable();
+      return;
+    }
+
+    // NOTE: For platforms without multihop support, we cannot emulate multihop
+    // and use port 53 at the same time. If the user has selected both options
+    // then let's choose multihop.
+    exitConfig.m_serverPort = exitServer.multihopPort();
+    exitConfig.m_serverIpv4AddrIn = entryServer.ipv4AddrIn();
+    exitConfig.m_serverIpv6AddrIn = entryServer.ipv6AddrIn();
+  }
+
+  m_activationQueue.append(exitConfig);
+  m_serverData.setEntryServerPublicKey(
+      m_activationQueue.first().m_serverPublicKey);
+  m_serverData.setExitServerPublicKey(
+      m_activationQueue.last().m_serverPublicKey);
+
+  m_pingReceived = false;
+  m_pingCanary.start(m_activationQueue.first().m_serverIpv4AddrIn, "0.0.0.0/0");
+  logger.info() << "Canary Ping Started";
+  activateNext();
+}
+
+void ConnectionManager::clearConnectedTime() {
```

Comment:
Out of curiosity, why did you decide to reorder a bunch of the functions from how they were in `controller.cpp`? Totally fine, I'm just genuinely curious.

Reply:
I created the connection manager class and started moving functions over from the controller one by one and moved all dependencies for each function as needed so it ended up being a little different in ordering. If there's code that you think should be ordered differently let me know. In hindsight I could've just copied practically all of controller code over and worked backwards  🫠

Reply:
Some of the ordering feels a bit less intuitive - activate/activateNext for example, but I'm certainly not going to block this on it. Thanks.

Reply:
If it helps at all, the next issue that is in progress is moving activate/deactivate and the corresponding helper back into the controller so it will be a little more intuitive soon :)